test/middleware/enforce_session_cookie_test.js

```diff
@@ -0,0 +1,79 @@
+'use strict'
+
+// NPM dependencies
+const assert = require('assert')
+const sinon = require('sinon')
+const { expect } = require('chai')
+const nock = require('nock')
+const { validChargeDetails } = require('../fixtures/payment_fixtures')
+const enforceSessionCookie = require('../../app/middleware/enforce_session_cookie')
+
+const ANALYTICS_ERROR = {
+    analytics: {
+        analyticsId: 'Service unavailable',
+        type: 'Service unavailable',
+        paymentProvider: 'Service unavailable',
+        amount: '0.00'
+    }
+}
+
+describe('enforce session cookie test', () => {
+    const response = {
+        status: () => { },
+        render: () => { },
+        locals: {}
+    }
+    let status
+    let render
+    let next
+    const validRequest = {
+        params: { chargeId: 'foo' },
+        body: {},
+        method: 'GET',
+        frontend_state: { ch_foo: true },
+        headers: {}
+    }
+    const chargeId = 'foo'
+
+    beforeEach(() => {
+        status = sinon.stub(response, 'status')
+        render = sinon.stub(response, 'render')
+        next = sinon.spy()
+        nock.cleanAll()
+    })
+
+    afterEach(() => {
+        status.restore()
+        render.restore()
+    })
+
+    it('should call not found view if frontend_state cookie not present', () => {
+        enforceSessionCookie({ params: {}, body: {} }, response, next)
+        assert(status.calledWith(403))
+        assert(render.calledWith('errors/incorrect_state/session_expired', {
+            viewName: 'UNAUTHORISED',
+            analytics: ANALYTICS_ERROR.analytics
+        }))
+        expect(next.notCalled).to.be.true
+    })
+
+    it('should call next when frontend_state cookie contains chargeId', done => {
+        const chargeData = validChargeDetails().getPlain()
+        nock(process.env.CONNECTOR_HOST)
+            .get(`/v1/frontend/charges/${chargeId}`)
+            .reply(200, chargeData)
+        enforceSessionCookie(validRequest, response, next)
+
+        const testPromise = new Promise((resolve) => {
```

Comment:
This looks like how this was tested before so potentially we shouldn't prolong this any further by changing it here but this has a number of red flags: 
* Test result is based on an arbitrary timeout instead of a chain of events, this could be flakey if any part CI is slow and could provide false negatives
* What is actually being tested is particularly hard to follow

As this test isn't testing the interaction between the application and the backend that could be stubbed instead of using `nock` OR `next` could be a promise that's passed in and the test depends on the fulfilment of that instead of waiting for a timeout. Happy to chat about that separately.

Reply:
May be can fix this in a separate PR, as you said it was originally test like this. If we are good with that, are you able to approve the change.

We discuss the appropriate test in detail later.